numpy/core/src/multiarray/ctors.c

```diff
@@ -3264,17 +3264,20 @@ array_fromfile_binary(FILE *fp, PyArray_Descr *dtype, npy_intp num, size_t *nrea
         }
         num = numbytes / dtype->elsize;
     }
+    Py_INCREF(dtype);
     r = (PyArrayObject *)PyArray_NewFromDescr(&PyArray_Type,
                                               dtype,
                                               1, &num,
                                               NULL, NULL,
                                               0, NULL);
     if (r == NULL) {
-        return NULL;
+        goto fail;
     }
     NPY_BEGIN_ALLOW_THREADS;
     *nread = fread(PyArray_DATA(r), dtype->elsize, num, fp);
     NPY_END_ALLOW_THREADS;
+fail:
+    Py_DECREF(dtype);
```

Comment:
Is it possible that the lost reference only occurs when there is also another error? I suspect that: https://github.com/rainwoodman/numpy/blob/patch-1/numpy/core/src/multiarray/ctors.c#L3425 may be the culprit. since the `array_fromfile_binary` steals a reference (also on error).